node/src/actors/chain_manager/handlers.rs

```diff
@@ -469,35 +469,41 @@ impl Handler<AddBlocks> for ChainManager {
 
                                     // We must construct the second superblock in order to be able
                                     // to validate the votes for this superblock later
-                                    log::debug!("Will construct the second superblock during synchronization. Superblock index: {} Epoch {}", sync_target.superblock.checkpoint + 1, candidate_superblock_epoch);
+                                    log::debug!("Will construct the second superblock during synchronization. Superblock index: {} Epoch {}", sync_target.superblock.checkpoint + 1, candidate_superblock_checkpoint * superblock_period);
 
-                                    actix::fut::ok(candidate_superblock_epoch)
+                                    actix::fut::ok(candidate_superblock_checkpoint)
                                 }
                             })
-                            .and_then(move |candidate_superblock_epoch, act, ctx| {
-                                act.build_and_vote_candidate_superblock(ctx, candidate_superblock_epoch).map(move |_, _, _| candidate_superblock_epoch)
+                            .and_then(move |candidate_superblock_checkpoint, act, ctx| {
+                                if let Some(candidate_superblock_epoch) = act.superblock_candidate_is_needed(candidate_superblock_checkpoint, superblock_period) {
+                                    actix::fut::Either::A(act.build_and_vote_candidate_superblock(ctx, candidate_superblock_epoch).map(move |_, _, _| Some(candidate_superblock_epoch)))
+                                }
+                                else{
+                                    actix::fut::Either::B(actix::fut::ok(None))
+                                }
                             })
                             .and_then(move |candidate_superblock_epoch, act, ctx| {
-                                // Process remaining blocks
-                                let (batch_succeeded, num_processed_blocks) = act.process_blocks_batch(ctx, &sync_target, &remaining_blocks);
-                                if !batch_succeeded {
-                                    log::error!("Received invalid blocks batch...");
-                                    act.update_state_machine(StateMachine::WaitingConsensus);
-                                    act.sync_waiting_for_add_blocks_since = None;
-
-                                    return actix::fut::err(());
-                                }
-                                log_sync_progress(&sync_target, &remaining_blocks, num_processed_blocks, "SyncWithCandidate(remaining)");
+                                if let Some(candidate_superblock_epoch) = candidate_superblock_epoch {
+                                    // Process remaining blocks
```

Comment:
Not sure if this `if` is correct. When `superblock_candidate_is_needed` returns `None` it means that the candidate superblock has already been created. But the remaining blocks are the ones after that superblock, so I think they should always be processed. Because if we already had this blocks we wouldn't have received them.

Reply:
The only think inside the `if` should be:

    act.move_chain_state_forward(superblock_index);

Because we don't need to save the chain state if we already have this superblock saved.